Fatras/src/Digitization/PlanarSurfaceMask.cpp

```diff
@@ -0,0 +1,276 @@
+    checkIntersection(
+        intersections,
+        intersector.intersectSegment(s0, s1, segment[0], sDir, true), sLength);
+  }
+  return maskAndReturn(intersections, segment, firstInside);
+}
+
+Acts::Result<ActsFatras::PlanarSurfaceMask::Segment2D>
+ActsFatras::PlanarSurfaceMask::radialMask(const Acts::RadialBounds& rBounds,
+                                          const Segment2D& segment,
+                                          const Segment2D& polarSegment,
+                                          bool firstInside) const {
+  double rMin = rBounds.get(Acts::RadialBounds::eMinR);
+  double rMax = rBounds.get(Acts::RadialBounds::eMaxR);
+  double hPhi = rBounds.get(Acts::RadialBounds::eHalfPhiSector);
+  double aPhi = rBounds.get(Acts::RadialBounds::eAveragePhi);
+
+  std::array<double, 2> radii = {rMin, rMax};
+  std::array<double, 2> phii = {aPhi - hPhi, aPhi + hPhi};
+
+  std::vector<Acts::Intersection2D> intersections;
+  Acts::Vector2D sVector(segment[1] - segment[0]);
+  Acts::Vector2D sDir = sVector.normalized();
+  double sLength = sVector.norm();
+
+  double sR = polarSegment[0][Acts::eBoundLoc0];
+  double eR = polarSegment[1][Acts::eBoundLoc0];
+  double sPhi = polarSegment[0][Acts::eBoundLoc1];
+  double ePhi = polarSegment[1][Acts::eBoundLoc1];
+
+  // Helper method to intersect phi boundaries
+  auto intersectPhiLine = [&](double phi) -> void {
+    Acts::Vector2D s0(rMin * std::cos(phi), rMin * std::sin(phi));
+    Acts::Vector2D s1(rMax * std::cos(phi), rMax * std::sin(phi));
+    checkIntersection(
+        intersections,
+        intersector.intersectSegment(s0, s1, segment[0], sDir, true), sLength);
+  };
+
+  // Helper method to intersect radial full boundaries
+  auto intersectCircle = [&](double r) -> void {
+    auto cIntersections = intersector.intersectCircle(r, segment[0], sDir);
+    for (const auto& intersection : cIntersections) {
+      checkIntersection(intersections, intersection, sLength);
+    }
+  };
+
+  // Intersect phi lines
+  if ((M_PI - hPhi) > Acts::s_epsilon) {
+    if (sPhi < phii[0] or ePhi < phii[0]) {
+      intersectPhiLine(phii[0]);
+    }
+    if (sPhi > phii[1] or ePhi > phii[1]) {
+      intersectPhiLine(phii[1]);
+    }
+    // Intersect radial segments
+    if (sR < radii[0] or eR < radii[0]) {
+      checkIntersection(intersections,
+                        intersector.intersectCircleSegment(
+                            radii[0], phii[0], phii[1], segment[0], sDir),
+                        sLength);
+    }
+    if (sR > radii[1] or eR > radii[1]) {
+      checkIntersection(intersections,
+                        intersector.intersectCircleSegment(
+                            radii[1], phii[0], phii[1], segment[0], sDir),
+                        sLength);
+    }
+  } else {
+    // Full radial set
+    // Intersect radial segments
+    if (sR < radii[0] or eR < radii[0]) {
+      intersectCircle(radii[0]);
+    }
+    if (sR > radii[1] or eR > radii[1]) {
```

Comment:
Hi @XiaocongAi - you are good. Indeed, this is the case that could happen, BUT ... This masker will operate on hits that are on surface and the reason why they could be outside of the mask is purely due to the incident angle and the charge drift. 

Both makes it practically impossible to create a case you describe.

As > 99 % of all cases are safe with both inside and the masker will not even operate, I do not want to penalise the 99% good cases with extra checks, that are pathological.

I will clarify this in the description.

Reply:
Yeah, from the realistic point of view, we are indeed safe.

Reply:
This is what I added:

```c++
  /// @note If both end points of the segment are inside, the segment
  /// is not clipped/masked, even if it would cross a surface boundary.
  /// Examples for those would be non-covex polygons or segments on a
  /// radial bound, where the radial boundary is crossed. Such segments
  /// do not occur in Digitization, as the hit has to be inside the 
  /// surface bounds to start with.
```

Reply:
Great! :+1: